docs/proposals/volume-idempotency.md

```diff
@@ -0,0 +1,229 @@
+
+### Non-idempotent volumes
+
+The hostPath volume does not implement an idempotency check because its `SetUp` method is a no-op.
+It should be possible to represent that a volume does not have any idempotency concerns.
+
+### Refreshable volume plugins
+
+Some volume plugins are designed to refresh, like the downward API volume plugin.  In order to
+decouple the implementation of refresh and SetUp logic, there should be a Refresh() method that
+plugins can implement and proper machinery to ensure that only plugins that want this need to
+implement it.
+
+### Volume plugins that wrap one another
+
+Some volume plugins wrap others:
+
+1.  Secrets, downward API, git repo are based on EmptyDir
+2.  PV claim plugin instantiates instances of the underlying volume plugin
+
+In order to correctly externalize idempotency management, the Kubelet must handle nested calls to
+`SetUp`; plugins must be able to invoke the externalized flow correctly for plugins they use
+transitively.
+
+## Ownership management and idempotency
+
+Ownership management of volumes must be idempotent for non-refreshable volumes.  It should run once
+and only once after `Setup` has run correctly.  If a volume that uses the 'check for a mountpoint'
+type of idempotency check, and the volume must be remounted, the ownership management should be
+reapplied.
+
+### Ownership management and refreshable volumes
+
+For volume plugins that are refreshable, ownership management must be reapplied if the refresh logic
+makes changes to the state of the volume.  For example, if data changes in the downward API volume,
+the ownership management must be re-applied to that volume.
+
+## Proposed implementation
+
+Summarizing the requirements for correct idempotency management:
+
+1.  It must be possible to express that a volume plugin has no idempotency check
+2.  It must be possible to express that a volume plugin uses either a mountpoint check or marker
+    file
+3.  Marker file volumes must have the marker file created after `SetUp` runs successfully
+4.  Refresh logic for refreshable plugins must be factored out of `SetUp`, and be run only if
+    `SetUp` has run successfully.
+5.  Ownership management must run once for each time `Setup` is invoked or refresh logic changes
+    the state of a volume
+6.  Volume plugins must be able to trigger the externalized Setup flow to support correct composition
+    of plugins
+
+The volume `Attributes` structure is the appropriate place to store information the properties we
+discuss here:
+
+```go
+package volume
+
+type Attributes struct {
+	// other fields omitted
+
+	IdempotencyStrategy IdempotencyStrategy
+	Refreshable          bool
+}
+
+type IdempotencyStrategyType string
+
+const (
+    IdempotencyStrategyNone       IdempotencyStrategyType = "none"
+    IdempotencyStrategyMountpoint IdempotencyStrategyType = "mountPoint"
+    IdempotencyStrategyMarkerFile IdempotencyStrategyType = "markerFile"
+)
```

Comment:
hmm.. this seems too expressive. Don't we just need decide whether a volume needs owndership management applied or not ? ie a boolean should suffice.